### PR TITLE
Board with player movement

### PR DESCRIPTION
**Pull request by shaylasry**
Created new Board and Player scriptable objects.
Created Player generator and player controller.
Implemented 2 movement approaches:
1. Character controller alone (use arrows for movement by default)
2. Character controller with Unity Input system (use W, S, A, D, or left joystick for movement)

Both know to handle floor collisions so the player will stop when reaching board boundaries.
Please notice that I used the 1.0f as ray distance for Physics.Raycast because the player Y coordinate is 1 and the board is 0.

### CODE REVIEW

File: Assets/Animations/PlayerInput.cs
Comment (itamarbiton) [inline Assets/Animations/PlayerInput.cs:1]: Why is this in `Animations` folder? Create a new folder under `Assets` name it `Input` and put it there

File: Assets/Animations/PlayerInput.cs.meta
Comment (itamarbiton) [inline Assets/Animations/PlayerInput.cs.meta:1]: Move to `Input` folder

File: Assets/Animations/PlayerInput.inputactions
Comment (itamarbiton) [inline Assets/Animations/PlayerInput.inputactions:1]: Move to `Input` folder

File: Assets/Animations/PlayerInput.inputactions.meta
Comment (itamarbiton) [inline Assets/Animations/PlayerInput.inputactions.meta:1]: Move to `Input` folder

File: Assets/Data/Player.asset
Comment (itamarbiton) [inline Assets/Data/Player.asset:1]: the only thing in this Scriptable object is the player's prefab? Why do we need this? can't we just use the prefab itself?

File: Assets/ScriptableObjects/Player.cs
Comment (itamarbiton) [inline Assets/ScriptableObjects/Player.cs:1]: Again, the only thing in this scriptable object is the player's prefab? Is this really needed?

File: Assets/Scripts/GameScene/IPlayerController.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    {
+    }
+
+    private void HandleCollision()
+    {
+    }
+
+    private void HandlePhysics()
+    {
+    }
```
Comment (itamarbiton) [inline Assets/Scripts/GameScene/IPlayerController.cs]: why do we need the brackets here? it's just a function declaration inside an interface, remove the brackets...

File: Assets/Scripts/GameScene/PlayerController.cs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+using UnityEngine;
+using UnityEngine.InputSystem;
+
+public class PlayerController : MonoBehaviour, IPlayerController
+{
+    [SerializeField] private Board board;
+    private PlayerInput _playerInput;
+    private CharacterController _characterController;
+    
+
```
Comment (itamarbiton) [inline Assets/Scripts/GameScene/PlayerController.cs]: why the double spacing? 1 line spacing is enough

File: Assets/Scripts/GameScene/PlayerController.cs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    [SerializeField] private Board board;
+    private PlayerInput _playerInput;
+    private CharacterController _characterController;
+    
+    
+    private Vector2 _movementInput;
+    private Vector3 _playerMovement;
+    private float _playerSpeed = 2.0f;
+
+
```
Comment (itamarbiton) [inline Assets/Scripts/GameScene/PlayerController.cs]: again remove the double spacing

File: Assets/Scripts/GameScene/PlayerController.cs [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+public class PlayerController : MonoBehaviour, IPlayerController
+{
+    [SerializeField] private Board board;
+    private PlayerInput _playerInput;
+    private CharacterController _characterController;
+    
+    
+    private Vector2 _movementInput;
+    private Vector3 _playerMovement;
+    private float _playerSpeed = 2.0f;
```
Comment (itamarbiton) [inline Assets/Scripts/GameScene/PlayerController.cs]: this should be configurable from outside of the script, make it `[SerializeField] private...`

File: Assets/Scripts/GameScene/PlayerController.cs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    
+    private void Awake()
+    {
+        InitPlayerInput();
+        _characterController = GetComponent<CharacterController>();
+    }
+    
+
+ 
+
```
Comment (itamarbiton) [inline Assets/Scripts/GameScene/PlayerController.cs]: too many line breaks, remove it, one is enough

File: Assets/Scripts/GameScene/PlayerController.cs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    
+    private bool _isMovementPressed;
+    [SerializeField] private float rotationFactorPerFrame = 1.0f;
+    
+    
+    private void Awake()
+    {
+        InitPlayerInput();
+        _characterController = GetComponent<CharacterController>();
```
Comment (itamarbiton) [inline Assets/Scripts/GameScene/PlayerController.cs]: Let's try to refrain from using `GetComponent` as it creates a dependency between the script and the GameObject it's attached to, make `_characterController` a `[SerializeField] private...`

File: Assets/Scripts/GameScene/PlayerController.cs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    }
+    
+
+ 
+
+    private void Update()
+    {
+        HandleMovementWithPlayerInput();
+        // HandleMovement();
+        // _characterController.Move(_playerMovement * Time.deltaTime * _playerSpeed);
```
Comment (itamarbiton) [inline Assets/Scripts/GameScene/PlayerController.cs]: why are these comments here? remove them

File: Assets/Scripts/GameScene/PlayerController.cs [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+    private void OnDisable()
+    {
+        _playerInput.PlayerContorls.Disable();
+    }
+
+    public Board GetGameBoard()
+    {
+        return AssetDatabase.LoadAssetAtPath<Board>("Assets/Data/Board.asset");
+    }
```
Comment (itamarbiton) [inline Assets/Scripts/GameScene/PlayerController.cs]: Let's try to refrain from doing this since it creates a dependency between the script and specific file in the project, remove this function and get the board's scriptable object as a `[SerializeField] private...`

File: Assets/Scripts/GameScene/PlayerController.cs [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+    
+    private void HandleCollision()
+    {
+        throw new System.NotImplementedException();
+    }
+
+    private void HandlePhysics()
+    {
+        throw new System.NotImplementedException();
+    }
```
Comment (itamarbiton) [inline Assets/Scripts/GameScene/PlayerController.cs]: why are these here? if they're not implemented remove them, no need for anything that's not implemented

Comment (itamarbiton) [inline Assets/Scripts/GameScene/PlayerController.cs]: what is this function supposed to do?

File: Assets/Scripts/GameScene/PlayerController.cs [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    
+    private void HandleMovementWithPlayerInput()
+    {
+        // Vector3 moveDirection = new Vector3(_playerMovement);
```
Comment (itamarbiton) [inline Assets/Scripts/GameScene/PlayerController.cs]: remove the comment

File: Assets/Scripts/GameScene/PlayerController.cs [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+        _playerMovement = transform.TransformDirection(_playerMovement);
+        Vector3 newPosition = transform.position + _playerMovement * Time.deltaTime * _playerSpeed;
+    
+      
+        RaycastHit hit;
+        float raycastDistance = 1f; 
+        if (Physics.Raycast(newPosition, Vector3.down, out hit, raycastDistance))
+        {
+            _characterController.Move(_playerMovement * Time.deltaTime * _playerSpeed);
```
Comment (itamarbiton) [inline Assets/Scripts/GameScene/PlayerController.cs]: why are we using Raycast here? Either use the board size, or explain why this is better

File: Assets/Scripts/GameScene/PlayerController.cs [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    
+      
+        RaycastHit hit;
+        float raycastDistance = 1f; 
+        if (Physics.Raycast(newPosition, Vector3.down, out hit, raycastDistance))
+        {
+            _characterController.Move(moveDirection * Time.deltaTime * _playerSpeed);
+        }
+        
+    }
```
Comment (itamarbiton) [inline Assets/Scripts/GameScene/PlayerController.cs]: what is the difference between this and `HandleMovementWithPlayerInput`? it's the same code duplicated! Either make it a function and call the function from both these methods or remove the one that we don't need

File: Assets/Scripts/GameScene/PlayerController.cs [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+        }
+    }
+    private void InitPlayerInput()
+    {
+        _playerInput = new PlayerInput();
+        
+        _playerInput.PlayerContorls.Movement.started += OnMovementInput;
+        _playerInput.PlayerContorls.Movement.performed += OnMovementInput;
+        _playerInput.PlayerContorls.Movement.canceled += OnMovementInput;
+    }
```
Comment (itamarbiton) [inline Assets/Scripts/GameScene/PlayerController.cs]: You're registering the receive events but your not removing the registration when the object is disabled/destroyed, that's bad practice, please register inside `OnEnable` and unregister in `OnDisable`

File: Assets/Scripts/GameScene/PlayerController.cs [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+    
+    private void OnMovementInput(InputAction.CallbackContext context)
+    {
+        _movementInput = context.ReadValue<Vector2>();
+
+        _playerMovement.x = _movementInput.x;
+        _playerMovement.y = 0.0f;
+        _playerMovement.z = _movementInput.y;
```
Comment (itamarbiton) [inline Assets/Scripts/GameScene/PlayerController.cs]: this can be done in 1 line of code, no need for 3... `_playerMovement = new Vector3(_movementInput.x, 0f, _movementInput.y);`

File: Assets/Scripts/GameScene/PlayerController.cs [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+        _isMovementPressed = _movementInput.x != 0 || _movementInput.y != 0;
```
Comment (itamarbiton) [inline Assets/Scripts/GameScene/PlayerController.cs]: this can be done in a much more simple way... `_movementInput != Vector2.Zero`

File: Assets/Scripts/GameScene/PlayerController.cs [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+
```
Comment (itamarbiton) [inline Assets/Scripts/GameScene/PlayerController.cs]: Remove this empty line

File: Assets/Scripts/GameScene/PlayerController.cs [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+        Vector3 playerDirection = new Vector3(_movementInput.x, 0.0f, _movementInput.y);
+        
+
+        Quaternion currentRotation = transform.rotation;
+        if (_isMovementPressed)
+        {
+            Quaternion targetRotation = Quaternion.LookRotation(playerDirection);
+            Quaternion.Slerp(currentRotation, targetRotation, rotationFactorPerFrame * Time.deltaTime);
+        }
+    }
```
Comment (itamarbiton) [inline Assets/Scripts/GameScene/PlayerController.cs]: If you can explain what's going on here I'll let you do it, but if not, please first understand what Quaternion.Slerp means before using it in this way

File: Assets/Scripts/GameScene/PlayerGenerator.cs [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+        GeneratePlayer();
+    }
+
+    private void GeneratePlayer()
+    {
+        Instantiate(player.prefab, new Vector3(0, 1, 0), Quaternion.identity);
+    }
+
+    
+
```
Comment (itamarbiton) [inline Assets/Scripts/GameScene/PlayerGenerator.cs]: Remove these empty lines

File: Assets/Scripts/GameScene/PlayerGenerator.cs [hunk 19]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+
+using UnityEngine;
+
+public class PlayerGenerator : MonoBehaviour
+{
+    [SerializeField] private Player player;
+    
+
```
Comment (itamarbiton) [inline Assets/Scripts/GameScene/PlayerGenerator.cs]: Remove this empty line

File: Assets/Scripts/GameScene/PlayerGenerator.cs [hunk 20]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+
+using UnityEngine;
+
+public class PlayerGenerator : MonoBehaviour
```
Comment (itamarbiton) [inline Assets/Scripts/GameScene/PlayerGenerator.cs]: Why do we need this script? What object uses it?

File: Logs/AssetImportWorker0-prev.log
Comment (itamarbiton) [inline Logs/AssetImportWorker0-prev.log:1]: don't push log files in the repo, add them to the .gitignore

**Review by itamarbiton — CHANGES_REQUESTED** [review]

File: Assets/Data/Player.asset
Comment (shaylasry) [reply Assets/Data/Player.asset:1]: For now, we can use the prefab. Made this so that if we want to make other related data like a username for the level selector we can use it.

File: Assets/Scripts/GameScene/PlayerController.cs (hunk 17 above)
Comment (shaylasry) [reply Assets/Scripts/GameScene/PlayerController.cs]: Transform is a Quaternion object.

we use the Slerp function to rotate from the current gameObject.transform to the new direction we go to according to the input we got. The time it takes to finish the rotation is: rotationFactorPerFrame * Time.deltaTime

I can figure out another way if you think this one is just complicated and we can do it better.

File: Assets/Scripts/GameScene/PlayerGenerator.cs (hunk 20 above)
Comment (shaylasry) [reply Assets/Scripts/GameScene/PlayerGenerator.cs]: If you mean the whole script file, Player Generator in Game Scene uses it the same as Board Generator uses BoardGenerator.cs

Do you want it to be in one merged file that generates both Board and Player?

**Comment by shaylasry** [comment]
1. resolved all the fixed issues, for non-resolved please see my comments.
2. almost forgot, according to Rider IDE a private serializeFIled and public filed should start with lower case and private only should start with '_'. 
I followed this pattern but I know you don't like the use of '_'. Should we check this once and for all to keep on the same pattern as known in the industry?
